dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ClientRateLimitHelper.java

```diff
@@ -0,0 +1,112 @@
+    /**
+     * The padding to keep from 0 remaining requests. If the &quot;X-RateLimit-Remaining&quut; header has a value
+     * smaller than this, wait for a rate limit reset before proceeding. This
+     */
+    private static final int REQUESTS_REMAINING_PADDING = 10;
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClientRateLimitHelper.class);
+    private final Duration maxWait;
+
+    /**
+     * Creates a helper, specifying how long to wait for a ratelimit reset.
+     * @param maxWait how long to wait for a reset
+     */
+    public ClientRateLimitHelper(Duration maxWait) {
+        this.maxWait = maxWait;
+    }
+
+    /**
+     * Checks if rate limit headers are present. If they are present, and the remaining limit is less than {@code BACK_OFF}, the method
+     * sleeps until the reset time is hit. Otherwise, does nothing.
+     *
+     * <p>If the wait for the reset time would be more than the <code>maxWait</code> parameter passed to the constructor, then it throws
+     * a <code>CustomWebApplicationException</code>.</p>
+     *
+     * @param headers the headers of an HTTP response
+     */
+    public void checkRateLimit(Map<String, List<String>> headers) {
+        getRemainingAndReset(headers).ifPresent(remainingAndReset -> {
+            if (remainingAndReset.remaining < REQUESTS_REMAINING_PADDING) {
+                LOG.info(headers.toString());
+                final Instant now = Instant.now();
+                LOG.info("Now is {}", now);
+                final Duration durationToReset = Duration.between(now, remainingAndReset.resetTime);
+                LOG.info("Duration is {}", durationToReset);
+                if (durationToReset.compareTo(maxWait) > 0) {
+                    throw new CustomWebApplicationException("Rate limits on an external service hit", HttpStatus.SC_INTERNAL_SERVER_ERROR);
+                }
+                if (!durationToReset.isNegative()) {
+                    try {
+                        LOG.info("Sleeping for {} seconds", durationToReset.toSeconds());
+                        Thread.sleep(durationToReset.toMillis());
+                    } catch (InterruptedException e) {
+                        LOG.error("Interrupted exception while waiting for reset limit", e);
+                    }
+                }
+            }
+        });
+    }
+
+    private Optional<RemainingAndReset> getRemainingAndReset(Map<String, List<String>> headers) {
+        final Map<String, List<String>> map = new CaseInsensitiveMap(headers);
+        final List<String> remaining = map.get(REMAINING_HEADER);
+        LOG.info("Remaining is {}", remaining);
+        if (remaining != null && !remaining.isEmpty()) {
+            final List<String> reset = map.get(RESET_HEADER);
+            if (reset != null && !reset.isEmpty()) {
+                final int remainingRequests = Integer.parseInt(remaining.get(0));
+                final int resetTime = Integer.parseInt(reset.get(0));
```

Comment:
If this code used a helper method to get the values out of the map, it might be easier to understand.